### PR TITLE
[WIP] px4_watchdog

### PR DESCRIPTION
**Pull request by dakejahl**
Not sure if there are any unintended consequences of creating this as a `template<class T>`.

Anyways, simple class interface to start a watchdog timer for a driver/module. The watchdog expiration callback needs to be defined by the class using it. The callback could involve resetting/rescheduling the driver or just killing it all together.

Addresses #11334

### CODE REVIEW

File: src/platforms/px4_watchdog.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+#include <perf/perf_counter.h>
+
+namespace px4
+{
+
+template<class T>
+class Watchdog
+{
+public:
+	Watchdog(unsigned interval, unsigned timeout) :
```
Comment (davids5) [inline src/platforms/px4_watchdog.h]: define units

File: src/platforms/px4_watchdog.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	{
+		hrt_call_after(&_call, delay, (hrt_callout)&T::watchdog_callback, this);
+	}
+
+	// Add additional time to the watchdog timer. This must be called every loop in order to kee the timer from expiring.
+	void kick()
+	{
+		perf_end(_kick_period);
+
+		hrt_abstime new_expiration = hrt_absolute_time() + _interval + _timeout;
```
Comment (davids5) [inline src/platforms/px4_watchdog.h]: What is the value in + _interval + _timeout;?
What it the thread here? Is it interruptible? 
Calculation will not be accurate w/o atomic operation. 
I am not convinced we should be using the hrt. But if we do I suggest extending the hrt to get a new deadline. (if that function in not there already)

File: src/platforms/px4_watchdog.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		perf_free(_kick_period);
+	}
+
+	// Start the watchdog timer with an optional  delay.
+	void start(hrt_abstime delay = 0)
+	{
+		hrt_call_after(&_call, delay, (hrt_callout)&T::watchdog_callback, this);
+	}
+
+	// Add additional time to the watchdog timer. This must be called every loop in order to kee the timer from expiring.
```
Comment (davids5) [inline src/platforms/px4_watchdog.h]: sp error

Comment (dakejahl) [reply src/platforms/px4_watchdog.h]: sp?

File: src/platforms/px4_watchdog.h (hunk 1 above)
Comment (dakejahl) [reply src/platforms/px4_watchdog.h]: thanks!

File: src/platforms/px4_watchdog.h (hunk 3 above)
Comment (davids5) [reply src/platforms/px4_watchdog.h]: 'kee ' spelling.

Comment (dakejahl) [reply src/platforms/px4_watchdog.h]: spelling :wink: thanks

**Comment by dakejahl** [comment]
> What is the facility of taking the timeout call back off an ISR? What would it do?

Maybe instead this watchdog should be a periodic item that runs in the HP `px4::wq`. Drivers that want to use this could register themselves with the watchdog and provide a callback function of their own that the watchdog can use to schedule reset/restart sequence for that driver. This assumes that these drivers are configured to be scheduled via a DRDY ISR callback which invokes `ScheduledWorkItem::ScheduleDelayed`. Does this sound correct to you @dagar?

> A facility that drivers call to check in. A short single service isr that checks for missed deadlines for all drivers then invokes schedules a long service (work call back) on the drives that missed.

Yeah! Should this be a `hrt_call_every`  or run in the `px4::wq`? 

> All this is subject to quantization issues.

Could you explain?

**Comment by davids5** [comment]
>>All this is subject to quantization issues.

>Could you explain?

It is why usleep is spec-ed to be no less than and rounded up
 See http://pubs.opengroup.org/onlinepubs/009695399/functions/usleep.html

Consider what would happen if sleep 500us on a system with 1 ms tick
System tick -------|-------|-------|-------|-------|-------|-------|-------|
a) usleep---------^   it would sleep 3 us.                         
b) usleep----------^ it would sleep 997us 
hece the round up so usleep(500) is really 2000us

Same applies on the periodic check-in relative to a single timer
checker -----------|-------|-------|-------|-------|-------|-------|-------|
a) checkin--------^--------^--------^--------^--------^--------^--------^
status.......................+...............-.................-................+                       
+ is early
- is late

**Comment by dakejahl** [comment]
> Same applies on the periodic check-in relative to a single timer

Oh okay, yes I understand this. I was thinking we'd look to make sure it checked in at least within 2x the expected interval. Does the rest of that I proposed sound reasonable? Was this what you had in mind?

**Comment by davids5** [comment]
> > What is the facility of taking the timeout call back off an ISR? What would it do?
> 
> Maybe instead this watchdog should be a periodic item that runs in the HP `px4::wq`. 

I think that may be unsafe on a wq and not deterministic. 

>Drivers that want to use this could register themselves with the watchdog and provide a callback function of their own that the watchdog can use to schedule reset/restart sequence for that driver. This assumes that these drivers are configured to be scheduled via a DRDY ISR callback which invokes `ScheduledWorkItem::ScheduleDelayed`. Does this sound correct to you @dagar?

I think there should be a SystemIntegrity module with its OWN task and possible with a few threads. It would monitor the failed check ins and also so do invariant state driven memory CRC verifications (poormans MPU) See https://github.com/PX4/Firmware/issues/6481#issuecomment-277450875
It could do the callouts on timeouts but the more I think out it. The HRT makes sense if the olny action is set a state in the driver to failed and post the semaphore. 

> 
> > A facility that drivers call to check in. A short single service isr that checks for missed deadlines for all drivers then invokes schedules a long service (work call back) on the drives that missed.
> 
> Yeah! Should this be a `hrt_call_every` or run in the `px4::wq`?

If we think of using the hrt as you have don, the hrt is being used as a retriggerable one-shot . What you have makes sense. I made a bad assumption in that the HRT would be firing a lot.  That was wrong.  So in reality the penalty of interrupt disable time is 1) Kick Path:The time in the critical sections in the hrt code to push the time out.   2) failed path:The time to set a flag and post the semaphore

Note: hrt_call_delay looks wrong to me I think there needs to be in a critical section (64 bit math on 32 bit machine)! See https://github.com/PX4/Firmware/issues/8099#issuecomment-459327995

**Comment by davids5** [comment]
> > Same applies on the periodic check-in relative to a single timer
> 
> Oh okay, yes I understand this. I was thinking we'd look to make sure it checked in at least within 2x the expected interval. Does the rest of that I proposed sound reasonable? Was this what you had in mind?

I think so: the WD is delayed 2 X the sensor interval.

**Comment by dakejahl** [comment]
> 1) Kick Path:The time in the critical sections in the hrt code to push the time out. 


> 2) failed path:The time to set a flag and post the semaphore

**Pull request closed without merging**